libr/bin/format/elf/elf.c

```diff
@@ -2627,150 +2635,276 @@ char *Elf_(r_bin_elf_get_rpath)(ELFOBJ *bin) {
+		diff = info->addr_rel + info->rel_size - g_offset - base_addr;
+		return diff;
+	} else if (info->addr_jmprel - base_addr <= g_offset && g_offset < info->addr_jmprel + info->jmprel_size - base_addr) {
+		diff = info->addr_jmprel + info->jmprel_size - g_offset - base_addr;
+		return diff;
 	}
-	bin->reloc_num = reloc_num;
-	ret = (RBinElfReloc*)calloc ((size_t)reloc_num + 1, sizeof (RBinElfReloc));
-	if (!ret) {
-		return NULL;
+
+	return offset;
+}
+
+static size_t populate_relocs_record_from_section(ELFOBJ *bin,
+	struct dynamic_relocation_section *info, RBinElfReloc *relocs, size_t pos) {
+	int res, rel_mode, i, j;
+	size_t size;
+
+	if (!bin->g_sections) {
+		return pos;
 	}
-	for (i = 0, rel = 0; !bin->g_sections[i].last && rel < reloc_num ; i++) {
-		bool is_rela = 0 == strncmp (bin->g_sections[i].name, ".rela.", strlen (".rela."));
-		bool is_rel  = 0 == strncmp (bin->g_sections[i].name, ".rel.",  strlen (".rel."));
-		if (!is_rela && !is_rel) {
+
+	for (i = 0; !bin->g_sections[i].last; i++) {
+		rel_mode = get_section_mode (bin, i);
+
+		if (!is_reloc_section (rel_mode) || bin->g_sections[i].size > bin->size || bin->g_sections[i].offset > bin->size) {
 			continue;
 		}
-		for (j = 0; j < bin->g_sections[i].size; j += res) {
-			if (bin->g_sections[i].size > bin->size) {
-				break;
-			}
-			if (bin->g_sections[i].offset > bin->size) {
-				break;
-			}
-			if (rel >= reloc_num) {
-				bprintf ("Internal error: ELF relocation buffer too small,"
-				         "please file a bug report.");
+
+		size = get_size_rel_mode(rel_mode);
+
+		for (j = get_next_not_analysed_offset (bin->g_sections[i].offset, 0, info, bin->baddr);
+			j < bin->g_sections[i].size;
+			j = get_next_not_analysed_offset (bin->g_sections[i].offset, j + size, info, bin->baddr)) {
+
+			res = read_reloc (bin, relocs + pos, rel_mode, bin->g_sections[i].offset + j);
+			if (res < 0) {
 				break;
 			}
-			if (!bin->is_rela) {
-				rela = is_rela? DT_RELA : DT_REL;
-			} else {
-				rela = bin->is_rela;
-			}
-			res = read_reloc (bin, &ret[rel], rela, bin->g_sections[i].offset + j);
-			if (j + res > bin->g_sections[i].size) {
-				bprintf ("malformed file, relocation entry #%u is partially beyond the end of section %u.\n", rel, i);
-			}
 			if (is_bin_etrel (bin)) {
-				if (bin->g_sections[i].info < bin->ehdr.e_shnum && bin->shdr) {
-					ret[rel].rva = bin->shdr[bin->g_sections[i].info].sh_offset + ret[rel].offset;
-					ret[rel].rva = Elf_(r_bin_elf_p2v) (bin, ret[rel].rva);
+				if (has_valid_section_header(bin, i)) {
+					relocs[pos].rva = bin->shdr[bin->g_sections[i].info].sh_offset + relocs[pos].offset;
+					relocs[pos].rva = Elf_(r_bin_elf_p2v) (bin, relocs[pos].rva);
 				} else {
-					ret[rel].rva = ret[rel].offset;
+					relocs[pos].rva = relocs[pos].offset;
 				}
 			} else {
-				ret[rel].rva = ret[rel].offset;
-				ret[rel].offset = Elf_(r_bin_elf_v2p) (bin, ret[rel].offset);
+				relocs[pos].rva = relocs[pos].offset;
```

Comment:
this code is the same as the one in `fix_rva_and_offset`. Maybe by changing a bit that function we can just call it here as well, instead of the whole `if(is_bin_etrel (bin)) ....`